cdap-common/src/main/java/io/cdap/cdap/common/service/ServiceDiscoverable.java

```diff
@@ -43,32 +45,41 @@
   private static final Set<ProgramType> USER_SERVICE_TYPES = Collections.unmodifiableSet(EnumSet.of(ProgramType.SERVICE,
                                                                                                     ProgramType.SPARK));
 
+  // mapping from the short discoverable component name to the program type
+  private static final Map<String, ProgramType> SHORT_NAME_TO_PROGRAM_TYPE_MAPPING = new HashMap<>();
+  static {
+    for (ProgramType userServiceType : USER_SERVICE_TYPES) {
+      SHORT_NAME_TO_PROGRAM_TYPE_MAPPING.put(userServiceType.getDiscoverableTypeName(), userServiceType);
+    }
+  }
+
   public static String getName(ProgramId programId) {
     return getName(programId.getNamespace(), programId.getApplication(), programId.getType(), programId.getProgram());
   }
 
   public static String getName(String namespaceId, String appId, ProgramType programType, String programName) {
-    if (!USER_SERVICE_TYPES.contains(programType)) {
-      throw new IllegalArgumentException("Program type should be one of " + USER_SERVICE_TYPES);
+    if (!SHORT_NAME_TO_PROGRAM_TYPE_MAPPING.containsValue(programType)) {
```

Comment:
Why need to change this? The `programType` should be in the `USER_SERVICE_TYPES`, right?

Reply:
we need a reverse mapping.
Lookup the ProgramType based on the short form.
We could have a switch statement instead...

Reply:
you are doing a `SHORT_NAME_TO_PROGRAM_TYPE_MAPPING.containsValue(programType)`, which is equivalent to doing `USER_SERVICE_TYPES.contains(programType)`, isn't it?

Reply:
Ya but I had to change the data structure from set to map for the usage on line 76.

Reply:
But you didn't delete the `USER_SERVICE_TYPES` constant, so why not just use it here?

Reply:
I thought I had deleted the map.
I reverted this change now.